File: modules/custom/lua/tutorial_npc_namevis.lua
-----------------------------------
-- Non-retail QoL: Hide tutorial NPC blue dot after beginner quest completion.
-- Requires custom/cpp/tutorial_npc_namevis.cpp for per-player namevis updates.
-----------------------------------
require('modules/module_utils')
local m = Module:new('tutorial_npc_namevis')

local tutorialNpcsByZone =
{
    [xi.zone.SOUTHERN_SAN_DORIA] = 17719618, -- Alaune
    [xi.zone.BASTOK_MARKETS]     = 17739939, -- Gulldago
    [xi.zone.WINDURST_WOODS]     = 17764600, -- Selele
}

local function hideTutorialMarkerIfComplete(player)
    if player:getVar('HQuest[Tutorial]Prog') < 12 then
        return
    end

    local npcId = tutorialNpcsByZone[player:getZoneID()]
    if npcId and xi.custom and xi.custom.sendNpcNamevisToPlayer then
        xi.custom.sendNpcNamevisToPlayer(player, npcId, 0)
    end
end

m:addOverride('xi.player.onGameIn', function(player, firstLogin, zoning)
    super(player, firstLogin, zoning)

    if zoning then
        hideTutorialMarkerIfComplete(player)
    end
end)

m:addOverride('InteractionGlobal.onEventFinish', function(player, csid, option, npc, fallbackFn)
    local result = super(player, csid, option, npc, fallbackFn)

    hideTutorialMarkerIfComplete(player)

    return result
end)

return m

File: modules/custom/cpp/tutorial_npc_namevis.cpp
/************************************************************************
 * Non-retail QoL: Per-player tutorial NPC quest-marker visibility.
 *
 * namevis is global per NPC; this helper sends a one-off entity update so
 * only the target player sees a different icon state. Enable in init.txt:
 *   custom/cpp/tutorial_npc_namevis.cpp
 ************************************************************************/

#include "map/entities/baseentity.h"
#include "map/entities/charentity.h"
#include "map/lua/lua_baseentity.h"
#include "map/packets/basic.h"
#include "map/utils/moduleutils.h"
#include "map/utils/zoneutils.h"

namespace
{

void SendNpcNamevisToPlayer(CLuaBaseEntity* PLuaPlayer, uint32 npcId, uint8 namevis)
{
    if (!PLuaPlayer)
    {
        return;
    }

    auto* PChar = dynamic_cast<CCharEntity*>(PLuaPlayer->GetBaseEntity());
    if (!PChar)
    {
        return;
    }

    auto* PEntity = zoneutils::GetEntity(npcId, TYPE_NPC);
    if (!PEntity)
    {
        return;
    }

    const uint8 saved = PEntity->namevis;
    PEntity->namevis  = namevis;
    PChar->updateEntityPacket(PEntity, ENTITY_UPDATE, UPDATE_HP);
    PEntity->namevis = saved;
}

} // namespace

class TutorialNpcNamevisModule : public CPPModule
{
    void OnInit() override
    {
        TracyZoneScoped;

        sol::table xiTable = lua["xi"];
        if (!xiTable.valid())
        {
            xiTable     = lua.create_table();
            lua["xi"]   = xiTable;
        }

        sol::table custom = xiTable["custom"];
        if (!custom.valid())
        {
            custom           = lua.create_table();
            xiTable["custom"] = custom;
        }

        custom.set_function("sendNpcNamevisToPlayer", &SendNpcNamevisToPlayer);
    }
};

REGISTER_CPP_MODULE(TutorialNpcNamevisModule);
